Add necessary subset of wide char APIs to minipal

AaronRobinsonMSFT opened the pull request:
Create the narrow subset of `WCHAR` (wide character) APIs that coreclr needs and abstract over them on all platforms.

~~This was developed on macOS and has not been built on any other platform.~~

Removing PAL tests at present is ill-advised since there are Win32 emulated APIs that rely on the PAL versions of the functions. The goal here though is the PAL implementation of most of the wide character APIs is no longer used directly by coreclr.

AaronRobinsonMSFT committed:
Replace wcschr and wcsncpy

AaronRobinsonMSFT committed:
Replace wcsrchr

AaronRobinsonMSFT committed:
Replace wcscmp and wcsncmp

AaronRobinsonMSFT committed:
Replace wcslen

AaronRobinsonMSFT committed:
Replace wcscpy

AaronRobinsonMSFT committed:
Replace wcscat

AaronRobinsonMSFT committed:
Remove use of wcspbrk

AaronRobinsonMSFT committed:
Remove wcsnlen

AaronRobinsonMSFT committed:
Replace wcsstr

AaronRobinsonMSFT committed:
Replace wcstod, wcstoul and _wcstoui64

AaronRobinsonMSFT committed:
Update minipal with UTF-16 functions

AaronRobinsonMSFT committed:
Remove function redefines from PAL header

jkotas commented on src/coreclr/vm/corelib.h:
```diff
@@ -846,7 +846,7 @@ DEFINE_METHOD(STRING,               CTORF_SBYTEPTR,         Ctor,
 DEFINE_METHOD(STRING,               CTORF_SBYTEPTR_START_LEN_ENCODING, Ctor,            SM_PtrSByt_Int_Int_Encoding_RetStr)
 DEFINE_METHOD(STRING,               INTERNAL_COPY,          InternalCopy,               SM_Str_IntPtr_Int_RetVoid)
-DEFINE_METHOD(STRING,               WCSLEN,                 wcslen,                     SM_PtrChar_RetInt)
+DEFINE_METHOD(STRING,               WCSLEN,                 dn_wcslen,                     SM_PtrChar_RetInt)
```
This is managed method. It should not be renamed.

AaronRobinsonMSFT replied:
Oh yep. I reverted the managed conversion. Missed this one.

AaronRobinsonMSFT committed:
Fix Windows build

New APIs should follow the Windows secure
pattern and supply destination length when
appropriate.

AaronRobinsonMSFT committed:
Remove need of PAL_wcslen from LTTNG
generator.

AaronRobinsonMSFT committed:
Remove specified calling conventions for x86

AaronRobinsonMSFT committed:
Additional cleanup

jkotas commented:
I would separate u16 suffix with underscore, e.g. `strlen_u16` or `wcslen_u16`.

AaronRobinsonMSFT commented:
> I would separate u16 suffix with underscore, e.g. `strlen_u16` or `wcslen_u16`.

I'm going with the `strlen_u16` style. It better aligns with the preferred standard APIs.

AaronRobinsonMSFT committed:
Change new API names based on feedback.

jkotas commented on src/coreclr/minipal/Windows/dn-u16.cpp:
```diff
@@ -0,0 +1,75 @@
+
+size_t strlen_u16(const WCHAR* str)
+{
+    if (str == NULL)
```
Why do we need this null check? The standard Windows wcslen does not have it, so we should not need it - no code can assume this null check today.

jkotas commented on src/coreclr/minipal/Windows/dn-u16.cpp:
```diff
@@ -0,0 +1,75 @@
+    return ::wcsncmp(str1, str2, count);
+}
+
+WCHAR* strcpy_u16(WCHAR* dst, size_t dstLen, const WCHAR* src)
```
Should we keep the `_s` suffix?

AaronRobinsonMSFT replied:
I would prefer not to. We own the abstraction here and the `_s` suffix is just weird. Our internal APIs should all be "safe"/"secure".

jkotas replied:
The less secure versions without `_s` suffix are still fine to use for some of the APIs.

I think we should have a simple decoder ring for these APIs: make them exactly same as the `char` version, with `u16` suffix and `char` changed to `char16_t`. If we start improving names or behaviors of these APIs, it will become muddy.

It may look better if the u16 prefix is first, like `u16_strcpy_s`. People out there tend to prefer `u16_` prefix: https://github.com/search?q=strlen_u16&type=code has 40 hits, https://github.com/search?q=u16_strlen&type=code has 924 hits.

AaronRobinsonMSFT replied:
> The less secure versions without _s suffix are still fine to use for some of the APIs.

Agreed, but the point here is an abstraction that we want to use internally. I still have to go through the codebase and update all the `_s` APIs too. I'd prefer to simply state the API name rather than perpetuate the historical artifact of where they came from.

> It may look better if the u16 prefix is first, like u16_strcpy_s. People out there tend to prefer u16_ prefix:

That is fair and I will apply that.

I still don't like the `_s` there. It makes it more convenient for someone to write the less secure one when they just shouldn't be used. I don't want a full projection of standard APIs just because someone doesn't want to deal with the secure API. I don't see the decoder ring issue at all. The APIs are "secure" by definition rather than mimicing the standard history.

jkotas commented on src/coreclr/minipal/Windows/dn-u16.cpp:
```diff
@@ -0,0 +1,75 @@
+
+#include <dn-u16.h>
+
+size_t strlen_u16(const WCHAR* str)
```
Can we use `char16_t` type instead of `WCHAR` in all this?

AaronRobinsonMSFT replied:
Not yet. That is the longer term plan. I want to keep with `WCHAR` until we remove the PAL. Once we start creating a proper abstraction layer (i.e., `minipal`) that change will occur.

AaronRobinsonMSFT committed:
Remove check for null. It isn't in the standard
and all current code preemptively makes the
check.

AaronRobinsonMSFT committed:
API rename suggestion

AaronRobinsonMSFT commented:
/cc @am11

jkotas reviewed (approved):
Thank you!

jkotas reviewed (approved):
Thank you!

jkotas reviewed (approved):
Thank you!

janvorli reviewed (approved):
LGTM, thank you!
